mesonbuild/compilers.py

```diff
@@ -1425,6 +1446,270 @@ def can_compile(self, filename):
+
+    def get_compile_only_args(self):
+        return ['-c']
+
+    def get_linker_output_args(self, target):
+        return ['-of', target]
+
+    def get_include_args(self, path, is_system):
+        return ['-I' + path]
+
+    def get_warn_args(self, level):
+        if level == '2':
+            return ['-wi']
+        else:
+            return ['-w']
+
+    def get_coverage_args(self):
+        return ['-cov']
+
+    def get_buildtype_linker_args(self, buildtype):
+        return []
+
+    def get_std_exe_link_args(self):
+        return []
+
+    def get_buildtype_args(self, buildtype):
+        return d_ldc_buildtype_args[buildtype]
+
+    def get_pic_args(self):
+        return ['-relocation-model=pic']
+
+    def _translate_args(self, args):
+        ldcargs = []
+        # Translate common arguments to flags this compiler can
+        # understand.
+        # The flags might have been added by pkg-config files,
+        # and are therefore out of the user's control.
+        for arg in args:
+            if arg == '-pthread':
+                continue
+            if arg.startswith('-Wl,'):
+                linkargs = arg[arg.index(',')+1:].split(',')
+                for la in linkargs:
+                    ldcargs.append('-L' + la.strip())
+                continue
+            elif arg.startswith('-l'):
+                # translate library link flag
+                ldcargs.append('-L' + arg)
+                continue
+            ldcargs.append(arg)
+
+        return ldcargs
+
+    def unix_link_flags_to_native(self, args):
+        return self._translate_args(args)
+
+    def unix_compile_flags_to_native(self, args):
+        return self._translate_args(args)
+
+class DmdDCompiler(DCompiler):
+    def __init__(self, exelist, version, is_cross):
+        DCompiler.__init__(self, exelist, version, is_cross)
+        self.id = 'dmd'
+
+    def get_dependency_gen_args(self, outtarget, outfile):
+        # LDC using the -deps flag returns a non-Makefile dependency-info file, which
+        # the backends can not use. So we disable this feature for now.
+        return []
+
+    def get_output_args(self, target):
+        return ['-of' + target]
+
+    def get_werror_args(self):
+        return ['-w']
+
+    def get_compile_only_args(self):
+        return ['-c']
+
+    def get_linker_output_args(self, target):
+        return ['-of' + target]
+
+    def get_include_args(self, path, is_system):
+        return ['-I' + path]
+
+    def get_warn_args(self, level):
+        return []
+
+    def get_coverage_args(self):
+        return ['-cov']
+
+    def get_buildtype_linker_args(self, buildtype):
+        return []
+
+    def get_std_exe_link_args(self):
+        return []
+
+    def get_buildtype_args(self, buildtype):
+        return d_dmd_buildtype_args[buildtype]
+
+    def get_std_shared_lib_link_args(self):
+        return ['-shared', '-defaultlib=libphobos2.so']
+
+    def _translate_args(self, args):
```

Comment:
There are differences in DMD/LDC compiler flags, so it might be needed in future to separate the translation code out...
But you're right, for now we could move this to the DCompiler class.